index.bs

```diff
@@ -2222,6 +2267,95 @@ To <dfn>parse trigger specs</dfn> given a [=map=] |map|, a [=moment=]
+    1. [=list/iterate|For each=] |value| of |values|:
+        1. If |value| is not a string, return an error.
+        1. If |value|'s [=string/length=] is 0 or is greater than [=max length per attribution scope=], return an error.
+        1. [=set/Append=] |value| to |result|.
+1. If |result| is not [=set/is empty|empty=]:
+    1. If |attributionScopeLimit| is null, return an error.
+1. If |result|'s [=set/size=] is greater than |attributionScopeLimit|, return an error.
+1. If |result|'s [=set/size=] is greater than [=max number of attribution scopes per source=], return an error.
+1. Return |result|.
+
+To <dfn>find sources with common destinations and reporting origin</dfn> given an [=attribution source=] |pendingSource|:
+1. Let |cache| be the user agent's [=attribution source cache=].
+1. Let |matchingSources| be an [=list/is empty|empty=] [=list=].
+1. [=set/iterate|For each=] |source| of |cache|:
+    1. Let |commonDestinations| be the [=set/intersection=] of |source|'s [=attribution source/attribution destinations=] and |pendingSource|'s [=attribution source/attribution destinations=].
+    1. If |commonDestinations| [=set/is empty=], [=iteration/continue=].
+    1. If |source|'s [=attribution source/reporting origin=] and |pendingSource|'s [=attribution source/reporting origin=] are not [=same origin=], [=iteration/continue=].
+    1. [=list/Append=] |source| to |matchingSources|.
+1. Return |matchingSources|.
+
+To <dfn>remove sources with different max number of event states if necessary</dfn> given an [=attribution source=] |pendingSource|:
+1. If |pendingSource|'s [=attribution source/max number of event states=] is null, return.
+1. [=list/iterate|For each=] |source| of [=find sources with common destinations and reporting origin=] with |pendingSource|:
+    1. If |source|'s [=attribution source/max number of event states=] is null or equals |pendingSource|'s [=attribution source/max number of event states=], [=iteration/continue=].
+    1. [=set/Remove=] |source| from the [=attribution source cache=].
+
+To <dfn>remove sources with lower attribution scope limit from cache if necessary</dfn> given an [=attribution source=] |pendingSource|:
+1. If |pendingSource|'s [=attribution source/attribution scope limit=] is null, return.
+1. [=list/iterate|For each=] |source| of [=find sources with common destinations and reporting origin=] with |pendingSource|:
+    1. If |source|'s [=attribution source/attribution scope limit=] is null or greater than or equal to |pendingSource|'s [=attribution source/attribution scope limit=], [=iteration/continue=].
+    1. [=set/Remove=] |source| from the [=attribution source cache=].
+
+To <dfn>select attribution scopes to remove</dfn> given a positive 32-bit integer |attributionScopeLimit|, a [=site=] |destination| and [=suitable origin=] |reportingOrigin|:
+1. Let |map| be an [=map/is empty|empty=] [=map=].
+1. [=set/iterate|For each=] |source| of the [=attribution source cache=]:
```

Comment:
Would it be easier to read with the following steps:
1. Find matching sources: same reporting origin, contains destination, non-empty attribution scopes.
2. Sort the sources based on source time in descending order.
3. Collect attribution scopes in descending order, until hitting attribution scope limit.
4. At this point, collect source identifiers if any of the source’s scopes is not in the final set.

Reply:
we will still need to break ties if some of the scopes share the same source time. We will need to sort the scopes anyways. I don't see that simplifies the logic much tbh. 

I'd prefer this as it matches with Rb implementation perfectly.

Reply:
Matching with Rb implementation is not a good reason for how to spec.

I still think this algorithm is over complicated with a map to maintain the last used time. Breaking ties is straightforward with a proper less than algorithm.

Reply:
Synced offline. Simplified the logic to not scan the cache twice.